File: apps/api/app/schemas/evidence.py
# API Schemas - Evidence/Provenance
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any, Literal
from datetime import datetime


class ConfidenceLabel(str, Enum):
    HIGH = "high"
    MEDIUM = "medium"
    LOW = "low"


class ConfidenceComponents(BaseModel):
    spatial_coverage: float = Field(ge=0, le=1)
    temporal_freshness: float = Field(ge=0, le=1)
    sample_density: float = Field(ge=0, le=1)
    measurement_quality: float = Field(ge=0, le=1)
    method_stability: float = Field(ge=0, le=1)


class ConfidenceScore(BaseModel):
    label: ConfidenceLabel
    score: float = Field(ge=0, le=1)
    components: ConfidenceComponents
    explanation: str
    limitations: List[str] = Field(default_factory=list)


class DataFreshness(BaseModel):
    latest_profile: str  # ISO 8601
    days_old: int = Field(ge=0)
    source: Literal["argo_realtime", "argo_delayed", "climatology", "reanalysis", "demo"]


class RegionInfo(BaseModel):
    type: str
    name: Optional[str] = None
    min_lat: Optional[float] = None
    max_lat: Optional[float] = None
    min_lon: Optional[float] = None
    max_lon: Optional[float] = None
    lat: Optional[float] = None
    lon: Optional[float] = None
    radius_km: Optional[float] = None


class DepthRangeInfo(BaseModel):
    min: float
    max: float


class TimeRangeInfo(BaseModel):
    start: str
    end: str


class QualityFiltersInfo(BaseModel):
    filters: List[str]
    description: str


class SourceIdentifiers(BaseModel):
    dataset: str
    snapshot: str
    doi: Optional[str] = None
    source_urls: List[str] = Field(default_factory=list)


class QueryStep(BaseModel):
    step: int
    tool: str
    params: Dict[str, Any]
    result_count: int
    duration_ms: Optional[int] = None


class EvidenceRecord(BaseModel):
    float_ids: List[int] = Field(default_factory=list)
    profile_count: int = Field(ge=0)
    observation_count: int = Field(ge=0)
    region: RegionInfo
    depth_range_m: Optional[DepthRangeInfo] = None
    time_range: TimeRangeInfo
    quality_filters: QualityFiltersInfo
    data_freshness: DataFreshness
    confidence: ConfidenceScore
    query_steps: List[QueryStep] = Field(default_factory=list)
    limitations: List[str] = Field(default_factory=list)
    source_identifiers: SourceIdentifiers
    verified: bool = False
    verification_errors: Optional[List[str]] = None


class NumericClaim(BaseModel):
    claim: str
    value: float | str
    unit: str
    claim_id: str
    source: Literal["measurement", "aggregation", "comparison", "projection", "climatology"]
    verified: bool


class VerificationResult(BaseModel):
    all_verified: bool
    claims: List[NumericClaim]
    failed_claims: List[NumericClaim]
    summary: str

File: packages/shared-types/src/evidence.ts
// Evidence/Provenance Types
// Proof-carrying numbers for ocean measurements

export type ConfidenceLabel = 'high' | 'medium' | 'low';

export interface ConfidenceComponents {
  spatial_coverage: number;      // 0-1
  temporal_freshness: number;    // 0-1
  sample_density: number;        // 0-1
  measurement_quality: number;   // 0-1
  method_stability: number;      // 0-1
}

export interface ConfidenceScore {
  label: ConfidenceLabel;
  score: number;                 // 0-1
  components: ConfidenceComponents;
  explanation: string;           // Human-readable
  limitations: string[];
}

export interface DataFreshness {
  latest_profile: string;        // ISO 8601
  days_old: number;
  source: 'argo_realtime' | 'argo_delayed' | 'climatology' | 'reanalysis' | 'demo';
}

export interface RegionInfo {
  type: string;
  name?: string;
  min_lat?: number;
  max_lat?: number;
  min_lon?: number;
  max_lon?: number;
  lat?: number;
  lon?: number;
  radius_km?: number;
}

export interface DepthRangeInfo {
  min: number;
  max: number;
}

export interface TimeRangeInfo {
  start: string;
  end: string;
}

export interface QualityFiltersInfo {
  filters: string[];
  description: string;
}

export interface SourceIdentifiers {
  dataset: string;
  snapshot: string;              // ISO 8601
  doi?: string;
  source_urls: string[];
}

export interface QueryStep {
  step: number;
  tool: string;
  params: Record<string, unknown>;
  result_count: number;
  duration_ms?: number;
}

export interface EvidenceRecord {
  float_ids: number[];
  profile_count: number;
  observation_count: number;
  region: RegionInfo;
  depth_range_m?: DepthRangeInfo;
  time_range: TimeRangeInfo;
  quality_filters: QualityFiltersInfo;
  data_freshness: DataFreshness;
  confidence: ConfidenceScore;
  query_steps: QueryStep[];
  limitations: string[];
  source_identifiers: SourceIdentifiers;
  verified: boolean;
  verification_errors?: string[];
}

export interface NumericClaim {
  claim: string;                 // Human-readable claim
  value: number | string;
  unit: string;
  claim_id: string;              // Links to profile/observation/float
  source: 'measurement' | 'aggregation' | 'comparison' | 'projection' | 'climatology';
  verified: boolean;
}

export interface VerificationResult {
  all_verified: boolean;
  claims: NumericClaim[];
  failed_claims: NumericClaim[];
  summary: string;
}